pyccel/codegen/printing/cwrappercode.py

```diff
@@ -0,0 +1,351 @@
+
+class CWrapperCodePrinter(CCodePrinter):
+    def __init__(self, parser, settings={}):
+        CCodePrinter.__init__(self, parser,settings)
+        self._cast_functions_dict = OrderedDict()
+        self._function_wrapper_names = dict()
+        self._global_names = set()
+
+    def get_new_name(self, used_names, requested_name):
+        if requested_name not in used_names:
+            used_names.add(requested_name)
+            return requested_name
+        else:
+            incremented_name, _ = create_incremented_string(used_names, prefix=requested_name)
+            return incremented_name
+
+    def find_in_dtype_registry(self, dtype, prec):
+        try :
+            return dtype_registry[(dtype, prec)]
+        except KeyError:
+            return CCodePrinter.find_in_dtype_registry(self, dtype, prec)
+
+    def get_cast_function(self, used_names, cast_type, from_variable, to_variable):
+        if cast_type in self._cast_functions_dict:
+            return self._cast_functions_dict[cast_type]
+
+        cast_function_name = self.get_new_name(used_names, cast_type)
+        cast_function_arg = [from_variable]
+        cast_function_result = [to_variable]
+        cast_function_local_vars = []
+
+        #switch case of cast_type
+        if cast_type == 'pyint_to_bool':
+            cast_function_body = [Assign(cast_function_result[0], Bool(cast_function_arg[0]))]
+
+        elif cast_type == 'bool_to_pyobj':
+            cast_function_body = [IfTernaryOperator((Bool(cast_function_arg[0]),
+                [AliasAssign(cast_function_result[0], Py_True)]),
+                (BooleanTrue(), [AliasAssign(cast_function_result[0], Py_False)]))]
+
+        elif cast_type == 'pycomplex_to_complex':
+            real_part = Variable(dtype = NativeReal(),
+                            name = self.get_new_name(used_names, 'real_part'))
+            imag_part = Variable(dtype = NativeReal(),
+                            name = self.get_new_name(used_names, 'imag_part'))
+            cast_function_local_vars += [real_part, imag_part]
+            cast_function_body = [Assign(real_part, FuncCall('PyComplex_RealAsDouble', cast_function_arg))]
+            cast_function_body += [Assign(imag_part, FuncCall('PyComplex_ImagAsDouble', cast_function_arg))]
+            cast_function_body += [Assign(cast_function_result[0], FuncCall('__builtin_complex', [real_part, imag_part]))]
+
+        elif cast_type == 'complex_to_pycomplex':
+            real_part = Variable(dtype = NativeReal(),
+                            name = self.get_new_name(used_names, "real_part"))
+            imag_part = Variable(dtype = NativeReal(),
+                            name = self.get_new_name(used_names, "imag_part"))
+            cast_function_local_vars += [real_part, imag_part]
+            cast_function_body = [Assign(real_part, FuncCall('__real', cast_function_arg))]
+            cast_function_body += [Assign(imag_part, FuncCall('__imag', cast_function_arg))]
```

Comment:
I cannot find a definition of `__real` and `__imag` online. Can you provide a reference?

I can find a reference to `__real__` and `__imag__` but these are gcc specific functions which we therefore should not use. Instead we should prefer `creal`, `cimag` (see https://en.wikipedia.org/wiki/C_mathematical_functions#complex.h)

Reply:
the difference  between __real __real__  is just in gcc version , and to use creal and cimag we need to include <complex.h>

Reply:
>the difference between __real __real__ is just in gcc version 

Do you have a reference for this? I get no useful results when searching for `__real` on google

Reply:
> to use creal and cimag we need to include <complex.h>

I think it is to be expected that we will need to use `<complex.h>` when handling complex numbers. It is not the only case where an extra include needs to be added (e.g. `<stdio.h>` for print)

Reply:
https://www.gnu.org/software/gnu-c-manual/gnu-c-manual.pdf     //  Lexical Elements -- Keywords
https://help.semmle.com/QL/learn-ql/cpp/introduce-libraries-cpp.html  I had another link I just need to find it .  


> > the difference between __real **real** is just in gcc version
> 
> Do you have a reference for this? I get no useful results when searching for `__real` on google

Reply:
> https://help.semmle.com/QL/learn-ql/cpp/introduce-libraries-cpp.html I had another link I just need to find it .

This link is fine. It has `__real` marked as a GNU extension (GNU are the makers of gcc). This means that this will not necessarily work with other compilers and should be avoided